src/widget.rs

```diff
@@ -30,57 +30,181 @@ use std::ops::{Deref, DerefMut};
+        std::any::type_name::<Self>()
+    }
 
-    /// Prepare for painting.
+    #[doc(hidden)]
+    /// Get the (abridged) type name of the widget for debugging purposes.
+    /// You should not override this method.
+    fn short_type_name(&self) -> &'static str {
+        let name = self.type_name();
+        name.split('<')
+            .next()
+            .unwrap_or(name)
+            .split("::")
+            .last()
+            .unwrap_or(name)
+    }
+
+    /*
+    #[doc(hidden)]
+    /// From the current data, get a best-effort description of the state of
+    /// this widget and its children for debugging purposes.
+    fn debug_state(&self) -> DebugState {
+        #![allow(unused_variables)]
+        DebugState {
+            display_name: self.short_type_name().to_string(),
+            ..Default::default()
+        }
+    }
+    */
+
+    /// Computes max intrinsic/preferred dimension of a widget on the provided axis.
     ///
-    /// This method is currently a bit of a hack. It's similar to the one in
-    /// Druid, which is for incremental repaint, but the primary purpose of
-    /// this one is virtualized scrolling.
+    /// Max intrinsic/preferred dimension is the dimension the widget could take, provided infinite
+    /// constraint on that axis.
     ///
-    /// The fact that `cx` is LayoutCx is just laziness, it should have its
-    /// own cx. And the main methods on that cx should be for region-based
-    /// invalidation.
-    #[allow(unused)]
-    fn prepare_paint(&mut self, cx: &mut LayoutCx, visible: Rect) {}
+    /// If axis == Axis::Horizontal, widget is being asked to calculate max intrinsic width.
+    /// If axis == Axis::Vertical, widget is being asked to calculate max intrinsic height.
+    ///
+    /// Box constraints must be honored in intrinsics computation.
+    ///
+    /// AspectRatioBox is an example where constraints are honored. If height is finite, max intrinsic
+    /// width is *height * ratio*.
+    /// Only when height is infinite, child's max intrinsic width is calculated.
+    ///
+    /// Intrinsic is a *could-be* value. It's the value a widget *could* have given infinite constraints.
+    /// This does not mean the value returned by layout() would be the same.
+    ///
+    /// This method **must** return a finite value.
+    fn compute_max_intrinsic(
+        &mut self,
+        axis: Axis,
+        ctx: &mut LayoutCx,
+        bc: &BoxConstraints,
+    ) -> f64 {
+        match axis {
+            Axis::Horizontal => self.layout(ctx, bc).width,
+            Axis::Vertical => self.layout(ctx, bc).height,
+        }
+    }
+}
 
-    fn paint(&mut self, cx: &mut PaintCx, builder: &mut SceneBuilder);
+/// An axis in visual space.
+///
+/// Most often used by widgets to describe
+/// the direction in which they grow as their number of children increases.
+/// Has some methods for manipulating geometry with respect to the axis.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Axis {
```

Comment:
Are these axes directional? I.e. does `Vertical` imply an origin at the top (or bottom) of the visual space, or is that direction independent of this?